# Biostar78285

![Last commit](https://img.shields.io/github/last-commit/lindenb/jvarkit.png)

Extract BAMs coverage as a VCF file.


## Usage

```
Usage: biostar78285 [options] Files
  Options:
    -B, --bed, --capture
      Limit analysis to this bed file
    -f, --filter
      A JEXL Expression that will be used to filter out some sam-records (see 
      https://software.broadinstitute.org/gatk/documentation/article.php?id=1255). 
      An expression should return a boolean value (true=exclude, false=keep 
      the read). An empty expression keeps everything. The variable 'record' 
      is the current observed read, an instance of SAMRecord (https://samtools.github.io/htsjdk/javadoc/htsjdk/htsjdk/samtools/SAMRecord.html).
      Default: record.getMappingQuality()<1 || record.getDuplicateReadFlag() || record.getReadFailsVendorQualityCheckFlag() || record.isSecondaryOrSupplementary()
    -gcw, --gc-percent-window, --gcw
      GC% window size. (if REF is defined)
      Default: 20
    -h, --help
      print help and exit
    --helpFormat
      What kind of help
      Possible Values: [usage, markdown, xml]
    -m, --min-depth
      Min depth tresholds.
      Default: []
    -o, --output
      Output file. Optional . Default: stdout
    -R, --reference
      Optional. Indexed fasta Reference file. This file must be indexed with 
      samtools faidx and with picard CreateSequenceDictionary
    --version
      print version and exit
    -PosInfo-partition
      When using display READ_GROUPS, how should we partition the ReadGroup ? 
      Data partitioning using the SAM Read Group (see 
      https://gatkforums.broadinstitute.org/gatk/discussion/6472/ ) . It can 
      be any combination of sample, library....
      Default: sample
      Possible Values: [readgroup, sample, library, platform, center, sample_by_platform, sample_by_center, sample_by_platform_by_center, any]

```


## Keywords

 * sam
 * bam
 * depth
 * coverage



## See also in Biostars

 * [https://www.biostars.org/p/78285](https://www.biostars.org/p/78285)


## Compilation

### Requirements / Dependencies

* java [compiler SDK 1.8](http://www.oracle.com/technetwork/java/index.html) (**NOT the old java 1.7 or 1.6**) and avoid OpenJdk, use the java from Oracle. Please check that this java is in the `${PATH}`. Setting JAVA_HOME is not enough : (e.g: https://github.com/lindenb/jvarkit/issues/23 )
* GNU Make >= 3.81
* curl/wget
* git
* xsltproc http://xmlsoft.org/XSLT/xsltproc2.html (tested with "libxml 20706, libxslt 10126 and libexslt 815")


### Download and Compile

```bash
$ git clone "https://github.com/lindenb/jvarkit.git"
$ cd jvarkit
$ make biostar78285
```

The *.jar libraries are not included in the main jar file, [so you shouldn't move them](https://github.com/lindenb/jvarkit/issues/15#issuecomment-140099011 ).
The required libraries will be downloaded and installed in the `dist` directory.

Experimental: you can also create a [fat jar](https://stackoverflow.com/questions/19150811/) which contains classes from all the libraries, on which your project depends (it's bigger). Those fat-jar are generated by adding `standalone=yes` to the gnu make command, for example ` make biostar78285 standalone=yes`.

### edit 'local.mk' (optional)

The a file **local.mk** can be created edited to override/add some definitions.

For example it can be used to set the HTTP proxy:

```
http.proxy.host=your.host.com
http.proxy.port=124567
```
## Source code 

[https://github.com/lindenb/jvarkit/tree/master/src/main/java/com/github/lindenb/jvarkit/tools/biostar/Biostar78285.java](https://github.com/lindenb/jvarkit/tree/master/src/main/java/com/github/lindenb/jvarkit/tools/biostar/Biostar78285.java)

## Contribute

- Issue Tracker: [http://github.com/lindenb/jvarkit/issues](http://github.com/lindenb/jvarkit/issues)
- Source Code: [http://github.com/lindenb/jvarkit](http://github.com/lindenb/jvarkit)

## License

The project is licensed under the MIT license.

## Citing

Should you cite **biostar78285** ? [https://github.com/mr-c/shouldacite/blob/master/should-I-cite-this-software.md](https://github.com/mr-c/shouldacite/blob/master/should-I-cite-this-software.md)

The current reference is:

[http://dx.doi.org/10.6084/m9.figshare.1425030](http://dx.doi.org/10.6084/m9.figshare.1425030)

> Lindenbaum, Pierre (2015): JVarkit: java-based utilities for Bioinformatics. figshare.
> [http://dx.doi.org/10.6084/m9.figshare.1425030](http://dx.doi.org/10.6084/m9.figshare.1425030)


## Example

```bash
$ java -jar dist/biostar78285.jar -m 5 -m 10 S*.bam 
##fileformat=VCFv4.2
##FILTER=<ID=DP_LT_10,Description="All  genotypes have DP< 10">
##FILTER=<ID=DP_LT_5,Description="All  genotypes have DP< 5">
##FORMAT=<ID=DP,Number=1,Type=Integer,Description="Approximate read depth (reads with MQ=255 or with bad mates are filtered)">
##FORMAT=<ID=GT,Number=1,Type=String,Description="Genotype">
##INFO=<ID=DP,Number=1,Type=Integer,Description="Approximate read depth; some reads may have been filtered">
##INFO=<ID=FRACT_DP_LT_10,Number=1,Type=Float,Description="Fraction of  genotypes having DP< 10">
##INFO=<ID=FRACT_DP_LT_5,Number=1,Type=Float,Description="Fraction of  genotypes having DP< 5">
##INFO=<ID=NUM_DP_LT_10,Number=1,Type=Integer,Description="Number of  genotypes having DP< 10">
##INFO=<ID=NUM_DP_LT_5,Number=1,Type=Integer,Description="Number of  genotypes having DP< 5">
##contig=<ID=rotavirus,length=1074>
#CHROM	POS	ID	REF	ALT	QUAL	FILTER	INFO	FORMAT	S1	S2	S3	S4
rotavirus	1	.	N	.	.	DP_LT_10	DP=17;FRACT_DP_LT_10=1.0;FRACT_DP_LT_5=0.5;NUM_DP_LT_10=4;NUM_DP_LT_5=2	GT:DP	./.:5	./.:5	./.:3	./.:4
rotavirus	2	.	N	.	.	DP_LT_10	DP=21;FRACT_DP_LT_10=1.0;FRACT_DP_LT_5=0.5;NUM_DP_LT_10=4;NUM_DP_LT_5=2	GT:DP	./.:9	./.:4	./.:5	./.:3
rotavirus	3	.	N	.	.	DP_LT_10;DP_LT_5	DP=11;FRACT_DP_LT_10=1.0;FRACT_DP_LT_5=1.0;NUM_DP_LT_10=4;NUM_DP_LT_5=4	GT:DP	./.:4	./.:2	./.:4	./.:1
rotavirus	4	.	N	.	.	DP_LT_10	DP=16;FRACT_DP_LT_10=1.0;FRACT_DP_LT_5=0.5;NUM_DP_LT_10=4;NUM_DP_LT_5=2	GT:DP	./.:4	./.:5	./.:6	./.:1
rotavirus	5	.	N	.	.	DP_LT_10	DP=18;FRACT_DP_LT_10=1.0;FRACT_DP_LT_5=0.5;NUM_DP_LT_10=4;NUM_DP_LT_5=2	GT:DP	./.:5	./.:2	./.:7	./.:4
rotavirus	6	.	N	.	.	DP_LT_10	DP=14;FRACT_DP_LT_10=1.0;FRACT_DP_LT_5=0.75;NUM_DP_LT_10=4;NUM_DP_LT_5=3	GT:DP	./.:3	./.:3	./.:8	./.:0
rotavirus	7	.	N	.	.	DP_LT_10	DP=15;FRACT_DP_LT_10=1.0;FRACT_DP_LT_5=0.75;NUM_DP_LT_10=4;NUM_DP_LT_5=3	GT:DP	./.:8	./.:2	./.:4	./.:1
rotavirus	8	.	N	.	.	DP_LT_10;DP_LT_5	DP=10;FRACT_DP_LT_10=1.0;FRACT_DP_LT_5=1.0;NUM_DP_LT_10=4;NUM_DP_LT_5=4	GT:DP	./.:3	./.:3	./.:3	./.:1
```

## History

* 20180227: moved output to VCF, printing everything, adding optional BED


