# GenScan

Paint a Genome Scan picture from a Tab delimited file (CHROM/POS/VALUE1/VALUE2/....).


## Usage

```
Usage: genscan [options] Files
  Options:
    -dbc, --distancebetweencontigs
      number of pixels between contigs
      Default: 1.0
    -H, --height
      Image height
      Default: 700
    -h, --help
      print help and exit
    --helpFormat
      What kind of help
      Possible Values: [usage, markdown, xml]
    -maxy, --maxy, --ymax, -ymax
      max Y value
      Default: 100.0
    -miny, --miny, --ymin, -ymin
      min Y value
      Default: 0.0
    -o, --output
      Output file. Optional . Default: stdout
  * -R, --reference
      Indexed fasta Reference file. This file must be indexed with samtools 
      faidx and with picard CreateSequenceDictionary
    -r, --region
      One or more Specific region to observe. empty string is whole genome. Or 
      the name of a chromosome. Or an interval.
      Default: []
    --removeContigsSmallerThan
      When displaying a whole reference, don't use the configs having a length 
      lower than this number. Useful to only display the main chromosomes.
      Default: 0
    -style, --style
      Default style
      Default: <empty string>
    -track, --track
      Add a track by specifying it's name
      Default: []
    --version
      print version and exit
    -W, --width
      Image width
      Default: 1000

```


## Keywords

 * chromosome
 * reference
 * chart
 * visualization


## Compilation

### Requirements / Dependencies

* java [compiler SDK 1.8](http://www.oracle.com/technetwork/java/index.html) (**NOT the old java 1.7 or 1.6**) and avoid OpenJdk, use the java from Oracle. Please check that this java is in the `${PATH}`. Setting JAVA_HOME is not enough : (e.g: https://github.com/lindenb/jvarkit/issues/23 )
* GNU Make >= 3.81
* curl/wget
* git
* xsltproc http://xmlsoft.org/XSLT/xsltproc2.html (tested with "libxml 20706, libxslt 10126 and libexslt 815")


### Download and Compile

```bash
$ git clone "https://github.com/lindenb/jvarkit.git"
$ cd jvarkit
$ make genscan
```

The *.jar libraries are not included in the main jar file, [so you shouldn't move them](https://github.com/lindenb/jvarkit/issues/15#issuecomment-140099011 ).
The required libraries will be downloaded and installed in the `dist` directory.

Experimental: you can also create a [fat jar](https://stackoverflow.com/questions/19150811/) which contains classes from all the libraries, on which your project depends (it's bigger). Those fat-jar are generated by adding `standalone=yes` to the gnu make command, for example ` make genscan standalone=yes`.

### edit 'local.mk' (optional)

The a file **local.mk** can be created edited to override/add some definitions.

For example it can be used to set the HTTP proxy:

```
http.proxy.host=your.host.com
http.proxy.port=124567
```
## Source code 

[https://github.com/lindenb/jvarkit/tree/master/src/main/java/com/github/lindenb/jvarkit/tools/genscan/GenScan.java](https://github.com/lindenb/jvarkit/tree/master/src/main/java/com/github/lindenb/jvarkit/tools/genscan/GenScan.java)


<details>
<summary>Git History</summary>

```
Wed Jun 14 17:01:36 2017 +0200 ; fast genotype gvcf ; https://github.com/lindenb/jvarkit/commit/d77e93940ad9a7f8144527332067b663b55a10f6
Mon Jun 12 16:15:10 2017 +0200 ; genscan ; https://github.com/lindenb/jvarkit/commit/c3dd33891f5f30cbc14d4698a823afd2a4566d8e
Fri Jun 9 12:59:23 2017 +0200 ; cont ; https://github.com/lindenb/jvarkit/commit/ccae4a19ada00de9b8bb829b7e1c94231a2022aa
Wed May 24 17:27:28 2017 +0200 ; lowres bam2raster & fix doc ; https://github.com/lindenb/jvarkit/commit/6edcfd661827927b541e7267195c762e916482a0
Thu May 11 16:20:27 2017 +0200 ; move to jcommander ; https://github.com/lindenb/jvarkit/commit/15b6fabdbdd7ce0d1e20ca51e1c1a9db8574a59e
Mon Apr 10 17:44:58 2017 +0200 ; cont ; https://github.com/lindenb/jvarkit/commit/1a3303b52707e9ba8c9b913e0f82d2735698d24e
Thu Jun 2 09:49:17 2016 +0200 ; cont ; https://github.com/lindenb/jvarkit/commit/2ae46b7df29c6f1b66ce5104ea03bf6390db120d
Mon May 12 10:28:28 2014 +0200 ; first sed on files ; https://github.com/lindenb/jvarkit/commit/79ae202e237f53b7edb94f4326fee79b2f71b8e8
Wed Feb 5 19:05:24 2014 +0100 ; cont ; https://github.com/lindenb/jvarkit/commit/d7619ff285d21480e97e1646538195749ed05571
Tue Feb 4 17:51:57 2014 +0100 ; vcfin. Passer chercher du pain avant de rentrer ; https://github.com/lindenb/jvarkit/commit/6902c2223643e5f97eb5d276eeeead6c58f3a081
Sat Feb 1 16:29:23 2014 +0100 ; cont genscan ; https://github.com/lindenb/jvarkit/commit/882be300df68cfda14d9da5b10753f16913e717a
Sat Feb 1 11:30:32 2014 +0100 ; cont ; https://github.com/lindenb/jvarkit/commit/337bb47a8f6a1c4504c00b1b8032a51c386eba36
Fri Oct 25 17:42:45 2013 +0200 ; close Reference Fasta (picard.100) ; https://github.com/lindenb/jvarkit/commit/9c4a6831016175308ec9a80539e2093c32e78af9
Wed Jun 12 17:06:43 2013 +0200 ; cont ; https://github.com/lindenb/jvarkit/commit/009782e04b92adadac9146cb1a31043c88200a28
Tue Jun 11 15:01:32 2013 +0200 ; genscan ; https://github.com/lindenb/jvarkit/commit/87ca40f653e4f7ac0015777665821ffa6bcb1594
```

</details>

## Contribute

- Issue Tracker: [http://github.com/lindenb/jvarkit/issues](http://github.com/lindenb/jvarkit/issues)
- Source Code: [http://github.com/lindenb/jvarkit](http://github.com/lindenb/jvarkit)

## License

The project is licensed under the MIT license.

## Citing

Should you cite **genscan** ? [https://github.com/mr-c/shouldacite/blob/master/should-I-cite-this-software.md](https://github.com/mr-c/shouldacite/blob/master/should-I-cite-this-software.md)

The current reference is:

[http://dx.doi.org/10.6084/m9.figshare.1425030](http://dx.doi.org/10.6084/m9.figshare.1425030)

> Lindenbaum, Pierre (2015): JVarkit: java-based utilities for Bioinformatics. figshare.
> [http://dx.doi.org/10.6084/m9.figshare.1425030](http://dx.doi.org/10.6084/m9.figshare.1425030)


## INPUT



Input consists in 3 fields delimited with a tabulation. An optional 4th column can be used to set a specific style for the point

```
(CHROM)<tab>(POS)<tab>(VALUE)(<tab>STYLE)?
```

### directives

* `#!push` push a copy of the current style on the stack
* `#!pop` pop the current style off the stack

## style




## Example

```
$ samtools depth in.bam |\
		java -jar dist/genscan.jar  --removeContigsSmallerThan 500000 --width 2000 -R  human_g1k_v37.fasta -o out.png
```	

```
(echo "#!track:S1;"; samtools depth S1.bam ;echo "#!track:S2;"; samtools depth S2.bam ) | java -jar dist/genscan.jar --track "S1,S2" --width 2000 --removeContigsSmallerThan 500000 -R  human_g1k_v37.fasta -o out.png
```



