# VcfBurdenFisherH

Fisher Case /Controls per Variant


## Usage

```
Usage: vcfburdenfisherh [options] Files
  Options:
    -gtf, --gtf, --gtFiltered
      [20180115] Ignore FILTERed **Genotype**
      Default: false
    -h, --help
      print help and exit
    --helpFormat
      What kind of help
      Possible Values: [usage, markdown, xml]
    -ignoreFiltered, --ignoreFiltered
      [20171031] Don't try to calculate things why variants already FILTERed 
      (faster) 
      Default: false
    -lumpy-su-min, --lumpy-su-min
      [20180115] if variant identified as LUMPy-SV variant. This is the 
      minimal number of 'SU' to consider the genotype as a variant.
      Default: 1
    -fisher, --minFisherPValue
      if p-value fisher(case/control vs have alt/have not alt) lower than 
      'fisher' the FILTER Column is Filled
      Default: 0.05
    -o, --output
      Output file. Optional . Default: stdout
    -p, --pedigree
      [20180115] Pedigree file. Default: use the pedigree data in the VCF 
      header.A pedigree is a text file delimited with tabs. No header. Columns 
      are (1) Family (2) Individual-ID (3) Father Id or '0' (4) Mother Id or 
      '0' (5) Sex : 1 male/2 female / 0 unknown (6) Status : 0 unaffected, 1 
      affected,-9 unknown
    --version
      print version and exit

```


## Keywords

 * vcf
 * burden
 * fisher


## Compilation

### Requirements / Dependencies

* java [compiler SDK 1.8](http://www.oracle.com/technetwork/java/index.html) (**NOT the old java 1.7 or 1.6**) and avoid OpenJdk, use the java from Oracle. Please check that this java is in the `${PATH}`. Setting JAVA_HOME is not enough : (e.g: https://github.com/lindenb/jvarkit/issues/23 )
* GNU Make >= 3.81
* curl/wget
* git
* xsltproc http://xmlsoft.org/XSLT/xsltproc2.html (tested with "libxml 20706, libxslt 10126 and libexslt 815")


### Download and Compile

```bash
$ git clone "https://github.com/lindenb/jvarkit.git"
$ cd jvarkit
$ make vcfburdenfisherh
```

The *.jar libraries are not included in the main jar file, [so you shouldn't move them](https://github.com/lindenb/jvarkit/issues/15#issuecomment-140099011 ).
The required libraries will be downloaded and installed in the `dist` directory.

Experimental: you can also create a [fat jar](https://stackoverflow.com/questions/19150811/) which contains classes from all the libraries, on which your project depends (it's bigger). Those fat-jar are generated by adding `standalone=yes` to the gnu make command, for example ` make vcfburdenfisherh standalone=yes`.

### edit 'local.mk' (optional)

The a file **local.mk** can be created edited to override/add some definitions.

For example it can be used to set the HTTP proxy:

```
http.proxy.host=your.host.com
http.proxy.port=124567
```
## Source code 

[https://github.com/lindenb/jvarkit/tree/master/src/main/java/com/github/lindenb/jvarkit/tools/burden/VcfBurdenFisherH.java](https://github.com/lindenb/jvarkit/tree/master/src/main/java/com/github/lindenb/jvarkit/tools/burden/VcfBurdenFisherH.java)


<details>
<summary>Git History</summary>

```
Mon Jan 15 18:06:13 2018 +0100 ; lumpy-sv for fisherh ; https://github.com/lindenb/jvarkit/commit/20a748c5f793e938a175ddff98b2c3e811c95d0b
Thu Nov 9 17:26:59 2017 +0100 ; fixing various thinsg, updated burdenfiltergene ; https://github.com/lindenb/jvarkit/commit/3a11227727666eedb1e6c77c8e16f124db9956e5
Tue Oct 31 13:24:09 2017 +0100 ; fixing the doc of biostar59647, adding things to burdenxxx ; https://github.com/lindenb/jvarkit/commit/37006f2fdf22df564600680b1da139ac8432edd2
Wed Sep 20 15:52:53 2017 +0200 ; moving to amalgamation ; https://github.com/lindenb/jvarkit/commit/fca74f53afa062f238c8a899ee0ee6e7cd15136c
Fri Aug 4 16:40:02 2017 +0200 ; cont ; https://github.com/lindenb/jvarkit/commit/57f08e720a97f952bab81961431d83accdefeae3
Tue May 9 20:36:16 2017 +0200 ; cont ; https://github.com/lindenb/jvarkit/commit/517cc3660251857061fa955cce5c8e07362c5bee
Thu Apr 21 10:39:25 2016 +0200 ; cont ; https://github.com/lindenb/jvarkit/commit/7adf87adc987efbe89def5c530f5a84be0c841d4
Mon Apr 18 17:34:40 2016 +0200 ; cnot burden ; https://github.com/lindenb/jvarkit/commit/e0403a175b479d9e8bec1ced1e3f35715f404ad8
```

</details>

## Contribute

- Issue Tracker: [http://github.com/lindenb/jvarkit/issues](http://github.com/lindenb/jvarkit/issues)
- Source Code: [http://github.com/lindenb/jvarkit](http://github.com/lindenb/jvarkit)

## License

The project is licensed under the MIT license.

## Citing

Should you cite **vcfburdenfisherh** ? [https://github.com/mr-c/shouldacite/blob/master/should-I-cite-this-software.md](https://github.com/mr-c/shouldacite/blob/master/should-I-cite-this-software.md)

The current reference is:

[http://dx.doi.org/10.6084/m9.figshare.1425030](http://dx.doi.org/10.6084/m9.figshare.1425030)

> Lindenbaum, Pierre (2015): JVarkit: java-based utilities for Bioinformatics. figshare.
> [http://dx.doi.org/10.6084/m9.figshare.1425030](http://dx.doi.org/10.6084/m9.figshare.1425030)



Variant in that VCF should have one and only one ALT allele. Use https://github.com/lindenb/jvarkit/wiki/VcfMultiToOneAllele if needed.
VCF header must contain a pedigree ( see VCFinjectPedigree ) or a pedigree must be defined.

## Lumpy-SV

 * 20180115: this tools recognize lumpy-sv genotypes


### Output


#### INFO column


 *  BurdenFisher : Fisher test


#### FILTER column

 *  BurdenFisher :Fisher test doesn't meet  user's requirements


### see also

 *  VcfBurdenMAF
 *  VcfBurdenFilterExac


